### ISSUE TITLE
PHP7 ubuntu error: ‘new_value_length’ undeclared

### ISSUE THREAD

**Issue by jaspart** [mitallast/damerau-levenshtein#1]
Hello mitallast,

First of all, thank you very much for proposing damerau-levenstein as a PHP shared library.
I am experiencing a compilation issue :
```
/damerau-levenshtein$ sudo make
/bin/bash /directory/damerau-levenshtein/libtool --mode=compile cc  -I. -I/directory/damerau-levenshtein -DPHP_ATOM_INC -I/directory/damerau-levenshtein/include -I/directory/damerau-levenshtein/main -I/directory/damerau-levenshtein -I/usr/include/php/20151012 -I/usr/include/php/20151012/main -I/usr/include/php/20151012/TSRM -I/usr/include/php/20151012/Zend -I/usr/include/php/20151012/ext -I/usr/include/php/20151012/ext/date/lib  -DHAVE_CONFIG_H  -g -O2   -c /directory/damerau-levenshtein/damerau.c -o damerau.lo 
libtool: compile:  cc -I. -I/directory/damerau-levenshtein -DPHP_ATOM_INC -I/directory/damerau-levenshtein/include -I/directory/damerau-levenshtein/main -I/directory/damerau-levenshtein -I/usr/include/php/20151012 -I/usr/include/php/20151012/main -I/usr/include/php/20151012/TSRM -I/usr/include/php/20151012/Zend -I/usr/include/php/20151012/ext -I/usr/include/php/20151012/ext/date/lib -DHAVE_CONFIG_H -g -O2 -c /directory/damerau-levenshtein/damerau.c  -fPIC -DPIC -o .libs/damerau.o
/directory/damerau-levenshtein/damerau.c: In function ‘OnUpdate_damerau_language’:
/directory/damerau-levenshtein/damerau.c:143:38: warning: passing argument 1 of ‘mbfl_name2no_language’ from incompatible pointer type [-Wincompatible-pointer-types]
  no_language = mbfl_name2no_language(new_value);
                                      ^
In file included from /usr/include/php/20151012/ext/mbstring/libmbfl/mbfl/mbfl_string.h:36:0,
                 from /directory/damerau-levenshtein/php_damerau.h:6,
                 from /directory/damerau-levenshtein/damerau.c:11:
/usr/include/php/20151012/ext/mbstring/libmbfl/mbfl/mbfl_language.h:83:38: note: expected ‘const char *’ but argument is of type ‘zend_string * {aka struct _zend_string *}’
 MBFLAPI extern enum mbfl_no_language mbfl_name2no_language(const char *name);
                                      ^
/directory/damerau-levenshtein/damerau.c: In function ‘OnUpdate_damerau_internal_encoding’:
/directory/damerau-levenshtein/damerau.c:157:70: error: ‘new_value_length’ undeclared (first use in this function)
         return _php_mb_ini_mbstring_internal_encoding_set(new_value, new_value_length TSRMLS_CC);
                                                                      ^
/directory/damerau-levenshtein/damerau.c:157:70: note: each undeclared identifier is reported only once for each function it appears in
/directory/damerau-levenshtein/damerau.c:157:59: warning: passing argument 1 of ‘_php_mb_ini_mbstring_internal_encoding_set’ from incompatible pointer type [-Wincompatible-pointer-types]
         return _php_mb_ini_mbstring_internal_encoding_set(new_value, new_value_length TSRMLS_CC);
                                                           ^
/directory/damerau-levenshtein/damerau.c:67:5: note: expected ‘const char *’ but argument is of type ‘zend_string * {aka struct _zend_string *}’
 int _php_mb_ini_mbstring_internal_encoding_set(const char *new_value, uint new_value_length TSRMLS_DC)
     ^
Makefile:193: recipe for target 'damerau.lo' failed
make: *** [damerau.lo] Error 1
```

I looked around a little, and it seems to be a PHP7 compatibility issue.
`error: ‘new_value_length’ undeclared`

I replaced the line 157 of damerau-levenshtein/damerau.c with strlen:
`return _php_mb_ini_mbstring_internal_encoding_set(new_value, (uint)strlen(new_value) TSRMLS_CC);`

But the make test fails:
```
PHP         : /usr/bin/php7.0 
PHP_SAPI    : cli
PHP_VERSION : 7.0.18-0ubuntu0.16.04.1
ZEND_VERSION: 3.0.0
PHP_OS      : Linux - Linux JUPITER 4.4.0-87-generic #110-Ubuntu SMP Tue Jul 18 12:55:35 UTC 2017 x86_64
INI actual  : /directory/damerau-levenshtein/tmp-php.ini
More .INIs  :   
CWD         : /directory/damerau-levenshtein
Extra dirs  : 
VALGRIND    : Not used
=====================================================================
TIME START 2017-07-29 15:41:46
=====================================================================
FAIL Bug #1 (Unicode support) [tests/damerau.phpt] 
FAIL Bug #1 (Memory leaks) [tests/damerau_1.phpt] 
=====================================================================
TIME END 2017-07-29 15:41:46

=====================================================================
TEST RESULT SUMMARY
---------------------------------------------------------------------
Exts skipped    :    0
Exts tested     :   14
---------------------------------------------------------------------

Number of tests :    2                 2
Tests skipped   :    0 (  0.0%) --------
Tests warned    :    0 (  0.0%) (  0.0%)
Tests failed    :    2 (100.0%) (100.0%)
Expected fail   :    0 (  0.0%) (  0.0%)
Tests passed    :    0 (  0.0%) (  0.0%)
---------------------------------------------------------------------
Time taken      :    0 seconds
=====================================================================

=====================================================================
FAILED TEST SUMMARY
---------------------------------------------------------------------
Bug #1 (Unicode support) [tests/damerau.phpt]
Bug #1 (Memory leaks) [tests/damerau_1.phpt]
=====================================================================
```
Would you have an idea?

For your information, here are the compilation logs:

```
/bin/bash /directory/damerau-levenshtein/libtool --mode=compile cc  -I. -I/directory/damerau-levenshtein -DPHP_ATOM_INC -I/directory/damerau-levenshtein/include -I/directory/damerau-levenshtein/main -I/directory/damerau-levenshtein -I/usr/include/php/20151012 -I/usr/include/php/20151012/main -I/usr/include/php/20151012/TSRM -I/usr/include/php/20151012/Zend -I/usr/include/php/20151012/ext -I/usr/include/php/20151012/ext/date/lib  -DHAVE_CONFIG_H  -g -O2   -c /directory/damerau-levenshtein/damerau.c -o damerau.lo 
libtool: compile:  cc -I. -I/directory/damerau-levenshtein -DPHP_ATOM_INC -I/directory/damerau-levenshtein/include -I/directory/damerau-levenshtein/main -I/directory/damerau-levenshtein -I/usr/include/php/20151012 -I/usr/include/php/20151012/main -I/usr/include/php/20151012/TSRM -I/usr/include/php/20151012/Zend -I/usr/include/php/20151012/ext -I/usr/include/php/20151012/ext/date/lib -DHAVE_CONFIG_H -g -O2 -c /directory/damerau-levenshtein/damerau.c  -fPIC -DPIC -o .libs/damerau.o
/directory/damerau-levenshtein/damerau.c: In function ‘OnUpdate_damerau_language’:
/directory/damerau-levenshtein/damerau.c:143:38: warning: passing argument 1 of ‘mbfl_name2no_language’ from incompatible pointer type [-Wincompatible-pointer-types]
  no_language = mbfl_name2no_language(new_value);
                                      ^
In file included from /usr/include/php/20151012/ext/mbstring/libmbfl/mbfl/mbfl_string.h:36:0,
                 from /directory/damerau-levenshtein/php_damerau.h:6,
                 from /directory/damerau-levenshtein/damerau.c:11:
/usr/include/php/20151012/ext/mbstring/libmbfl/mbfl/mbfl_language.h:83:38: note: expected ‘const char *’ but argument is of type ‘zend_string * {aka struct _zend_string *}’
 MBFLAPI extern enum mbfl_no_language mbfl_name2no_language(const char *name);
                                      ^
/directory/damerau-levenshtein/damerau.c: In function ‘OnUpdate_damerau_internal_encoding’:
/directory/damerau-levenshtein/damerau.c:157:83: warning: passing argument 1 of ‘strlen’ from incompatible pointer type [-Wincompatible-pointer-types]
         return _php_mb_ini_mbstring_internal_encoding_set(new_value, (uint)strlen(new_value) TSRMLS_CC);
                                                                                   ^
In file included from /usr/include/php/20151012/main/php_config.h:2522:0,
                 from /usr/include/php/20151012/main/php.h:34,
                 from /directory/damerau-levenshtein/damerau.c:7:
/usr/include/string.h:394:15: note: expected ‘const char *’ but argument is of type ‘zend_string * {aka struct _zend_string *}’
 extern size_t strlen (const char *__s)
               ^
/directory/damerau-levenshtein/damerau.c:157:59: warning: passing argument 1 of ‘_php_mb_ini_mbstring_internal_encoding_set’ from incompatible pointer type [-Wincompatible-pointer-types]
         return _php_mb_ini_mbstring_internal_encoding_set(new_value, (uint)strlen(new_value) TSRMLS_CC);
                                                           ^
/directory/damerau-levenshtein/damerau.c:67:5: note: expected ‘const char *’ but argument is of type ‘zend_string * {aka struct _zend_string *}’
 int _php_mb_ini_mbstring_internal_encoding_set(const char *new_value, uint new_value_length TSRMLS_DC)
     ^
/bin/bash /directory/damerau-levenshtein/libtool --mode=link cc -DPHP_ATOM_INC -I/directory/damerau-levenshtein/include -I/directory/damerau-levenshtein/main -I/directory/damerau-levenshtein -I/usr/include/php/20151012 -I/usr/include/php/20151012/main -I/usr/include/php/20151012/TSRM -I/usr/include/php/20151012/Zend -I/usr/include/php/20151012/ext -I/usr/include/php/20151012/ext/date/lib  -DHAVE_CONFIG_H  -g -O2   -o damerau.la -export-dynamic -avoid-version -prefer-pic -module -rpath /directory/damerau-levenshtein/modules  damerau.lo 
libtool: link: cc -shared  -fPIC -DPIC  .libs/damerau.o    -g -O2   -Wl,-soname -Wl,damerau.so -o .libs/damerau.so
libtool: link: ( cd ".libs" && rm -f "damerau.la" && ln -s "../damerau.la" "damerau.la" )
/bin/bash /directory/damerau-levenshtein/libtool --mode=install cp ./damerau.la /directory/damerau-levenshtein/modules
libtool: install: cp ./.libs/damerau.so /directory/damerau-levenshtein/modules/damerau.so
libtool: install: cp ./.libs/damerau.lai /directory/damerau-levenshtein/modules/damerau.la
libtool: finish: PATH="/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/snap/bin:/sbin" ldconfig -n /directory/damerau-levenshtein/modules
----------------------------------------------------------------------
Libraries have been installed in:
   /directory/damerau-levenshtein/modules

If you ever happen to want to link against installed libraries
in a given directory, LIBDIR, you must either use libtool, and
specify the full pathname of the library, or use the '-LLIBDIR'
flag during linking and do at least one of the following:
   - add LIBDIR to the 'LD_LIBRARY_PATH' environment variable
     during execution
   - add LIBDIR to the 'LD_RUN_PATH' environment variable
     during linking
   - use the '-Wl,-rpath -Wl,LIBDIR' linker flag
   - have your system administrator add LIBDIR to '/etc/ld.so.conf'

See any operating system documentation about shared libraries for
more information, such as the ld(1) and ld.so(8) manual pages.
----------------------------------------------------------------------

Build complete.
Don't forget to run 'make test'.
```

**Comment by mitallast** [comment]
Sorry, there's no support for php7 – I'm not use php last 4-5 years.

PHP7 has so many changes in internal api. Best solution - rewrite extension based on origin core levenshtein function, using mb_string api, and add damerau extension.

**Comment by jaspart** [comment]
Thank you very much for the advice.

**Comment by jaspart** [comment]
Hello Alexey,
just a little message to say that your advice has been implemented for both 5.6 and 7.1
https://github.com/jaspart/damerau
Have a great one
Arnaud
